### PR TITLE
Experiment with unboxed options on Dotty compiler

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

**Comment by nicolasstucki** [comment]
test performance please

File: library/src/dotty/uoption/package.scala [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    def get: A =
+      if (isEmpty) throw new NoSuchElementException("UNone.get")
+      else forceGet
+
+    @inline def map[B](f: A => B): UOption[B] =
+      if (isEmpty) self.asInstanceOf[UOption[B]]
+      else USome(f(forceGet))
+
+    @inline def flatMap[B](f: A => UOption[B]): UOption[B] =
+      if (isEmpty) UNone else f(this.get)
```
Comment (sjrd) [inline library/src/dotty/uoption/package.scala]: `forceGet`

**Comment by nicolasstucki** [comment]
test performance please

**Comment by nicolasstucki** [comment]
test performance please

**Comment by nicolasstucki** [comment]
test performance please

**Comment by nicolasstucki** [comment]
When developing this, the main source of `Option` came from `collection.Map`.
We should try unboxed options in the collection strawman and test the performance there.

The important files developed durring this PR are:
https://github.com/lampepfl/dotty/pull/3181/files#diff-be4a0762db35fb639c5c69d78870f0a4
https://github.com/lampepfl/dotty/pull/3181/files#diff-750b1435fe47081e0ed6d1df70f0a22d

**Pull request closed without merging**